src/linux/component.rs

```diff
@@ -1,160 +1,369 @@
-                let mut p_input = folder.to_path_buf();
-                let mut p_crit = folder.to_path_buf();
-                let mut p_max = folder.to_path_buf();
-
-                p_label.push(&format!("temp{}_label", key));
-                p_input.push(&format!("temp{}{}", key, val[found_input]));
-                p_max.push(&format!("temp{}_max", key));
-                p_crit.push(&format!("temp{}_crit", key));
-                if is_file(&p_input) {
-                    let label = get_file_line(p_label.as_path(), 10)
-                        .unwrap_or_else(|| format!("Component {}", key)) // needed for raspberry pi
-                        .replace('\n', "");
-                    let max = get_file_line(p_max.as_path(), 10).map(|max| {
-                        max.replace('\n', "").parse::<f32>().unwrap_or(100_000f32) / 1000f32
-                    });
-                    let crit = get_file_line(p_crit.as_path(), 10).map(|crit| {
-                        crit.replace('\n', "").parse::<f32>().unwrap_or(100_000f32) / 1000f32
-                    });
-                    components.push(Component::new(label, p_input, max, crit));
+            "input" => {
+                let temperature = get_temperature_from_file(&hwmon_file);
+                component.input_file = hwmon_file;
+                component.temperature = temperature;
+                // Max computed value taken form current temperatue if not possible to
+                // get it from `highest`.
+                if let None = component.max {
+                    component.max = temperature;
                 }
             }
+            "label" => {
+                component.label = get_file_line(&hwmon_file, 10).unwrap_or(String::new())
+            }
+            "highest" => {
+                component.max = get_temperature_from_file(&hwmon_file);
+                component.highest_file = hwmon_file;
+            },
+            "max" => component.threshold_max = get_temperature_from_file(&hwmon_file),
+            "min" => component.threshold_min = get_temperature_from_file(&hwmon_file),
+            "crit" => component.threshold_critical = get_temperature_from_file(&hwmon_file),
+            // Raspberry_pi weird comportement.
+            "" if component.label.is_empty() => component.label = format!("Component name {}", id),
+            // TODO maybe it's good to trace for being future aware of a change in hwmon temp?
+            _ => {}
+        }
+    };
+
+    let dir = read_dir(folder).ok()?;
+    let mut matchings: HashMap<u32, Component> = HashMap::with_capacity(10);
+    for entry in dir.flatten() {
+        let entry = entry.path();
+        let filename = entry.file_name().and_then(|x| x.to_str()).unwrap_or("");
+        if entry.is_dir() || !filename.starts_with("temp") {
+            continue;
         }
+
+        let (id, item) = filename.split_once('_')?;
+        let id = id[4..].parse::<u32>().ok()?;
```

Comment:
I prefer to be over cautious about indexing: please don't use `4..` or at least confirm that its length is greater than 5 before doing it.

Reply:
No problem its in theory checked by start_with(`temp`) did like it was before:

```
parts.next().map(|s| s[4..].parse::<u32>().ok())
```
I am ok with length check or use `get(..4)`, this WIP is a good way to avoid a problem.

```
        if entry.is_dir() || !filename.starts_with("temp") {
            continue;
        }
        
        let (id, item) = filename.split_once('_')?;
        assert!(id.len(5))
        let id = id[4..].parse::<u32>().ok()?;
```
Or
```
        let (id, item) = filename.split_once('_')?;
        assert!(id.len(5))
        let id = id.get(4..)?.parse::<u32>().ok()?;
```

Reply:
You're right, no need for an extra check then. Can you add a comment like `// no need to check for the length since we already checked that it starts with "temp" above` please?

Reply:
perfers an `assert!` it's more future proof and llvm might use it.